File: translate.mjs
import fs from 'fs';
import { translate } from '@vitalets/google-translate-api';

const delay = ms => new Promise(res => setTimeout(res, ms));

const translateText = async (text, targetLang, retries = 3) => {
    if (typeof text !== 'string' || !text.trim()) return text;
    for (let i = 0; i < retries; i++) {
        try {
            const res = await Promise.race([
                translate(text, { to: targetLang }),
                new Promise((_, reject) => setTimeout(() => reject(new Error("Timeout")), 5000))
            ]);
            return res.text;
        } catch (e) {
            console.error(`Attempt ${i+1} failed for '${text.substring(0, 20)}...':`, e.message);
            await delay(1000 * (i + 1));
        }
    }
    return text; // Return original if all retries fail
};

const translateObject = async (obj, targetLang, basePath) => {
    const result = Array.isArray(obj) ? [] : {};
    let count = 0;
    
    // Load existing progress if any
    if (fs.existsSync(basePath)) {
        try {
           const existing = JSON.parse(fs.readFileSync(basePath, 'utf8'));
           Object.assign(result, existing);
        } catch(e) {}
    }

    for (const [key, value] of Object.entries(obj)) {
        if (result[key] !== undefined && Object.keys(result[key] || {}).length > 0) {
            continue; // Skip already translated keys at top level
        }
        
        console.log(`Translating section: ${key}`);
        if (typeof value === 'object' && value !== null) {
             // For simplicity in this script, we recursively translate but don't save mid-section
             const sectionResult = Array.isArray(value) ? [] : {};
             for (const [subKey, subValue] of Object.entries(value)) {
                 if (typeof subValue === 'object' && subValue !== null) {
                     // 1 layer deeper
                     const deeperResult = {};
                     for (const [deepKey, deepValue] of Object.entries(subValue)) {
                          if (typeof deepValue === 'string') {
                              deeperResult[deepKey] = await translateText(deepValue, targetLang);
                              await delay(100);
                          } else { deeperResult[deepKey] = deepValue; }
                     }
                     sectionResult[subKey] = deeperResult;
                 } else if (typeof subValue === 'string') {
                     sectionResult[subKey] = await translateText(subValue, targetLang);
                     await delay(100);
                 } else {
                     sectionResult[subKey] = subValue;
                 }
             }
             result[key] = sectionResult;
        } else if (typeof value === 'string') {
            result[key] = await translateText(value, targetLang);
            await delay(100);
        } else {
            result[key] = value;
        }
        
        // Save progress after every top-level key
        fs.writeFileSync(basePath, JSON.stringify(result, null, 2));
    }
    return result;
};

const main = async () => {
    const raw = fs.readFileSync('messages/en.json', 'utf8');
    const enData = JSON.parse(raw);
    
    // Process one language at a time, completely
    const langs = [
        'hi', 'pa', 'mr', 'bn', 'te', 'ta',
        'as', 'gu', 'kn', 'ml', 'or', 'ur',
        'sa', 'sd', 'ne', 'mai', 'doi', 'gom', 
        'sat', 'ks', 'mni'
    ];
    
    for (const lang of langs) {
        console.log(`\n=== Starting translation for ${lang} ===`);
        const targetFile = `messages/${lang}.json`;
        try {
            await translateObject(enData, lang, targetFile);
            console.log(`Finished translating ${lang}.json`);
        } catch (err) {
            console.error(`Failed on ${lang}:`, err.message);
        }
    }
};

main();

File: translate.py
import json
import urllib.request
import urllib.parse
import os
from typing import Dict, Any

def translate_text(text, target_lang):
    if not isinstance(text, str):
        return text
    if not text.strip():
        return text

    # Simple logic using a free translation API or google translate URL hack (not recommended for production, but works for script)
    url = f"https://translate.googleapis.com/translate_a/single?client=gtx&sl=en&tl={target_lang}&dt=t&q={urllib.parse.quote(text)}"
    try:
        req = urllib.request.Request(url, headers={'User-Agent': 'Mozilla/5.0'})
        with urllib.request.urlopen(req) as response:
            data = json.loads(response.read().decode())
            return "".join([x[0] for x in data[0]])
    except Exception as e:
        print(f"Error translating '{text}': {e}")
        return text

def translate_dict(d: Dict[str, Any], target_lang: str) -> Dict[str, Any]:
    translated: Dict[str, Any] = {}
    for k, v in d.items():
        if isinstance(v, dict):
            translated[k] = translate_dict(v, target_lang)
        elif isinstance(v, str):
            translated[k] = translate_text(v, target_lang)
        else:
            translated[k] = v
    return translated

def main():
    base_dir = r"c:\Users\HP\Desktop\ProjectFolder\krishimitra\messages"
    with open(os.path.join(base_dir, "en.json"), "r", encoding="utf-8") as f:
        en_data = json.load(f)

    langs = [
        "hi", "pa", "mr", "bn", "te", "ta",
        "as", "gu", "kn", "ml", "or", "ur",
        "sa", "sd", "ne", "mai", "doi", "gom", 
        "sat", "ks", "mni"
    ]
    for lang in langs:
        print(f"Translating to {lang}...")
        translated_data = translate_dict(en_data, lang)
        with open(os.path.join(base_dir, f"{lang}.json"), "w", encoding="utf-8") as f:
            json.dump(translated_data, f, ensure_ascii=False, indent=2)
        print(f"Finished {lang}.json")

if __name__ == "__main__":
    main()
